staging/src/k8s.io/apiextensions-apiserver/pkg/apiserver/schema/cel/validation_test.go

```diff
@@ -0,0 +1,1012 @@
+					},
+					"spec": map[string]interface{}{
+						"field1": "a",
+					},
+				},
+			},
+			schema: objectTypePtr(map[string]schema.Structural{
+				"embedded": {
+					Generic: schema.Generic{Type: "object"},
+					Extensions: schema.Extensions{
+						XEmbeddedResource: true,
+					},
+					Properties: map[string]schema.Structural{
+						"kind":       stringType,
+						"apiVersion": stringType,
+						"metadata": *objectTypePtr(map[string]schema.Structural{
+							"name":         stringType,
+							"generateName": stringType,
+						}),
+					},
+				},
+			}),
+			valid: []string{
+				"embedded.kind == 'Pod'",
+				"embedded.apiVersion == 'v1'",
+				"embedded.metadata.name == 'foo'",
+				"embedded.metadata.generateName == 'pickItForMe'",
+			},
+			// only field declared in the schema can be field selected in CEL expressions
+			errors: map[string]string{
+				"has(embedded.metadata.namespace)": "undefined field 'namespace'",
+				"has(embedded.spec)":               "undefined field 'spec'",
+			},
+		},
+		{name: "embedded object with properties",
+			obj: map[string]interface{}{
+				"embedded": map[string]interface{}{
+					"apiVersion": "v1",
+					"kind":       "Pod",
+					"metadata": map[string]interface{}{
+						"name":         "foo",
+						"generateName": "pickItForMe",
+						"namespace":    "xyz",
+					},
+					"spec": map[string]interface{}{
+						"field1": "a",
+					},
+				},
+			},
+			schema: objectTypePtr(map[string]schema.Structural{
+				"embedded": {
+					Generic: schema.Generic{Type: "object"},
+					Extensions: schema.Extensions{
+						XEmbeddedResource: true,
+					},
+					Properties: map[string]schema.Structural{
+						"kind":       stringType,
+						"apiVersion": stringType,
+						"spec": *objectTypePtr(map[string]schema.Structural{
+							"field1": stringType,
+						}),
+					},
+				},
+			}),
+			valid: []string{
+				"embedded.spec.field1 == 'a'",
+			},
+			// only field declared in the schema can be field selected in CEL expressions
+			errors: map[string]string{
+				"has(embedded.metadata)": "undefined field 'metadata'",
+			},
+		},
+		{name: "embedded with prune unknown",
+			obj: map[string]interface{}{
+				"embedded": map[string]interface{}{
+					"apiVersion": "v1",
+					"kind":       "Pod",
+					"metadata": map[string]interface{}{
+						"name":         "foo",
+						"generateName": "pickItForMe",
+						"namespace":    "xyz",
+					},
+					"spec": map[string]interface{}{
+						"field1": "a",
+					},
+				},
+			},
+			schema: objectTypePtr(map[string]schema.Structural{
+				"embedded": {
+					Extensions: schema.Extensions{
+						XPreserveUnknownFields: true,
+						XEmbeddedResource:      true,
+					},
+				},
+			}),
+			errors: map[string]string{
+				// we only provide information in the schema to validation rules
+				// without type=object, we don't provide access to the embedded resource
+				"has(self.embedded)": "undefined field 'embedded'",
```

Comment:
Shouldn't not at least `self.embedded` be available if `XEmbeddedResource` is true?

Reply:
This was an invalid test case. We require type=object on embedded resources. (I tested this on a running cluster).